iothub_client/src/iothub_client_diagnostic.c

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <math.h>
+#include "azure_c_shared_utility/optimize_size.h"
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/buffer_.h"
+
+#include "iothub_client_diagnostic.h"
+
+static const int BASE_36 = 36;
+
+static char* get_current_time_utc(char* timeBuffer, int bufferLen, const char* formatString)
+{
+    time_t epochTime;
+    struct tm* utcTime;
+    epochTime = time(NULL);
+    utcTime = gmtime(&epochTime);
+    strftime(timeBuffer, bufferLen, formatString, utcTime);
+    return timeBuffer;
+}
+
+static char get_base36_char(unsigned char value)
+{
+    if (value <= 9)
+    {
+        return '0' + value;
+    }
+    else
+    {
+        return 'a' + value - 10;
+    }
+}
+
+static char* generate_eight_random_characters(char *randomString)
+{
+    char* randomStringPos = randomString;
+    for (int i = 0; i < 4; ++i)
+    {
+        int rawRandom = rand();
+        int first = rawRandom % BASE_36;
+        int second = rawRandom / BASE_36 % BASE_36;
+        *randomStringPos++ = get_base36_char((unsigned char)first);
+        *randomStringPos++ = get_base36_char((unsigned char)second);
+    }
+    *randomStringPos = 0;
+
+    return randomString;
+}
+
+static bool should_add_diagnostic_info(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting)
+{
+    if (diagSetting->diagSamplingPercentage > 0)
+    {
+        if (diagSetting->currentMessageNumber == UINT32_MAX)
+        {
+            diagSetting->currentMessageNumber %= diagSetting->diagSamplingPercentage;
+        }
+        ++diagSetting->currentMessageNumber;
+
+        double number = diagSetting->currentMessageNumber;
+        double percentage = diagSetting->diagSamplingPercentage;
+        return (floor((number - 2) * percentage / 100.0) < floor((number - 1) * percentage / 100.0));
```

Comment:
Do you need to divide by 100.0 on both sides?

Reply:
yes, it is used to make sure sampling messages with diagnostic properties are uniform among all messages, e.g. if diagSamplingPercentage is equal to 20, the messages flow should be :
D N N N N D N N N N D...
D means message with diagnostic properties while N means normal message